### PR TITLE
Bug 1641910: Trying to set audit_log_exclude_accounts crashes server

### PR DESCRIPTION
**Pull request by gl-sergei**
The root cause of this bug is that logic around filtering variables rely
heavily on the fact that CHECK and UPDATE functions are always called.
This is not true when variables passed via defaults file or set as
command line options.

The fix is add special handling for filtering variables on startup.

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1477/

### CODE REVIEW

File: mysql-test/t/audit_log_acc.test [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Bug 1641910: Trying to set audit_log_exclude_accounts crashes server.
+
+call mtr.add_suppression("Plugin 'audit_log' init function returned error.");
+call mtr.add_suppression("Plugin 'audit_log' registration as a AUDIT failed.");
+
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
```
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_acc.test]: Here and everywhere please use include/restart_mysqld.inc, it supports restart parameters.

File: mysql-test/t/audit_log_acc.test [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Bug 1641910: Trying to set audit_log_exclude_accounts crashes server.
```
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_acc.test]: Use one of the existing test cases?

File: plugin/audit_log/audit_log.c [hunk 3]
```diff
@@ -811,7 +811,53 @@ int audit_log_plugin_init(void *arg MY_ATTRIBUTE((unused)))
 
   audit_log_filter_init();
 
-  return(0);
+  if (audit_log_exclude_accounts != NULL && audit_log_include_accounts != NULL)
```
Comment (laurynas-biveinis) [inline plugin/audit_log/audit_log.c:814]: IMHO this added code belongs to audit_log_filter_init, up to you to consider

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: plugin/audit_log/audit_log.c (hunk 3 above)
Comment (gl-sergei) [reply plugin/audit_log/audit_log.c:814]: audit_log_filer_init located in separate module and doesn't know about plugin variables. This code is aimed to validate plugin variables.

File: mysql-test/t/audit_log_acc.test (hunk 2 above)
Comment (gl-sergei) [reply mysql-test/t/audit_log_acc.test]: This test case does a lot of server restarts to test how different variables combinations are tested on startup. We don't have similar test cases currently.

Comment (gl-sergei) [reply mysql-test/t/audit_log_acc.test]: the name is bad... maybe name it audit_log_bug1641910?

File: mysql-test/t/audit_log_acc.test (hunk 1 above)
Comment (gl-sergei) [reply mysql-test/t/audit_log_acc.test]: done

File: mysql-test/t/audit_log_acc.test (hunk 2 above)
Comment (laurynas-biveinis) [reply mysql-test/t/audit_log_acc.test]: audit_log_startup ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by gl-sergei** [comment]
Renamed test case to `audit_log_startup`.
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1489/

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**